compiler/ilgen/IlBuilder.cpp

```diff
@@ -1565,6 +1565,122 @@ IlBuilder::AtomicAdd(TR::IlValue * baseAddress, TR::IlValue * value)
+ *    +----------------------------------------+    +-------------------------------+    +-------------------------+
+ *                      |                                          |                                 |
+ *                      |                                          |                                 |
+ *                      |------------------------------------------+---------------------------------+
+ *                      |
+ *                      v             
+ *              +----------------+     
+ *              | block_$merge   |     
+ *              | ============== |   
+ *              |      ...       |
+ *              +----------------+   
+ *
+ * \endverbatim 
+ *
+ * \structure & \param value
+ *     tstart
+ *       |
+ *       ----persistentFailure // This is a permanent failure, try atomic way to do it instead
+ *       |
+ *       ----transientFailure // Temporary failure, user can retry 
+ *       |
+ *       ----transaction // Success, use general(non-atomic) way to do it
+ *                |
+ *                ---- non-atomic operations
+ *                |
+ *                ---- ...
+ *                |
+ *                ---- tend
+ */
+void
+IlBuilder::Transaction(TR::IlBuilder **persistentFailureBuilder, TR::IlBuilder **transientFailureBuilder, TR::IlBuilder **transactionBuilder)
+    {   
+    //ILB_REPLAY("%s->TransactionBegin(%s, %s, %s);", REPLAY_BUILDER(this), REPLAY_BUILDER(persistentFailureBuilder), REPLAY_BUILDER(transientFailureBuilder), REPLAY_BUILDER(transactionBuilder)); 
+    TraceIL("IlBuilder[ %p ]::transactionBegin %p, %p, %p, %p)\n", this, *persistentFailureBuilder, *transientFailureBuilder, *transactionBuilder);
+   
+    appendBlock();
+
+    TR::Block *mergeBlock = emptyBlock();
+    *persistentFailureBuilder = createBuilderIfNeeded(*persistentFailureBuilder);
+    *transientFailureBuilder = createBuilderIfNeeded(*transientFailureBuilder);
+    *transactionBuilder = createBuilderIfNeeded(*transactionBuilder);
+
+    TR::Node *persistentFailureNode = TR::Node::create(TR::branch, 0, (*persistentFailureBuilder)->getEntry()->getEntry());
+    TR::Node *transientFailureNode = TR::Node::create(TR::branch, 0, (*transientFailureBuilder)->getEntry()->getEntry());
+    TR::Node *transactionNode = TR::Node::create(TR::branch, 0, (*transactionBuilder)->getEntry()->getEntry());
+
+    TR::Node *tStartNode = TR::Node::create(TR::tstart, 3, persistentFailureNode, transientFailureNode, transactionNode);   
+    tStartNode->setSymbolReference(comp()->getSymRefTab()->findOrCreateTransactionEntrySymbolRef(comp()->getMethodSymbol()));
+    
+    genTreeTop(tStartNode);
+
+    //connecting the block having tstart with persistentFailure's and transaction's blocks 
+    cfg()->addEdge(_currentBlock, (*persistentFailureBuilder)->getEntry());
+    cfg()->addEdge(_currentBlock, (*transactionBuilder)->getEntry());
```

Comment:
Since tstart does not actually fall-through, I would prefer you to add explicit control flow edges to all three blocks and then call appendNoFallThroughBlock(). It may feel strange, and it may end up doing the same thing, but I feel it more correctly models what's happening for the block containing that tstart node.